### ISSUE TITLE
Roll cycle possibly not picked up by tailer

### ISSUE THREAD

**Issue by jheusser** [OpenHFT/Chronicle-Queue#301]
Using Chronicle Queue 4.5.14, we have a long running tailer that outlives appender processes across roll cycles. So e.g. Monday morning the tailer starts and is still alive on Tuesday, while the appender processes start again on Tuesday morning after being shutdown Monday evening before midnight. Once the appender starts writing data to a new roll cycle (Tuesday) it seems like the tailer does not always catch that new data being written.

I have attempted to create a test replicating this behaviour: https://gist.github.com/jheusser/28621f722f0afa6137e80ef0e0afc276

If you uncomment line 115 where the appender writes data on day 1 the tailer picks up all data and the test passes, without it it's not reading anything and the test fails (with 0 documents read).

**Comment by RobAustin** [comment]
thanks - we will take a look, its likely to be a while before we will get a chance to look at this, when we do we will post, comments here.

**Comment by RobAustin** [comment]
I've created a simple test doing something similar, my test below is using test_secondly, which uses a roll cycle of just 1 second. The trailer is created and exists across all the seconds. 

``` java
    @Test
    public void testLongLivingTailerAppenderReAcquiredEachSecond() throws Exception {

        final Path dir = Files.createTempDirectory("demo");
        final RollCycles rollCycle = RollCycles.TEST_SECONDLY;

        // write first message
        try (ChronicleQueue queue = ChronicleQueueBuilder
                .single(dir.toString())
                .rollCycle(rollCycle).build()) {

            final ExcerptTailer tailer = queue.createTailer();

            for (int i = 0; i < 5; i++) {

                try (final DocumentContext dc = queue.acquireAppender().writingDocument()) {
                    dc.wire().write("some").int32(i);
                }

                try (final DocumentContext dc = tailer.readingDocument()) {
                    Assert.assertEquals(i, dc.wire().read("some").int32());
                }

                Thread.sleep(1000);
            }
        }
    }
````

**Comment by RobAustin** [comment]
I'll take a look what you have in https://gist.github.com/jheusser/28621f722f0afa6137e80ef0e0afc276 to see if I can find anything strange.

**Comment by jheusser** [comment]
Thanks @RobAustin. I've only ever observed the tailer missing events when the appender didn't produce any data for a roll cycle and then restarts on a new roll cylce. As described, Friday the appender stops but the tailer does not, Monday the appender produces new data and the tailer does not pick up the new cycle. My test produces the following output for me with no calls to `onReleased` or `onAcquired`:
```
java.lang.AssertionError: 
Expected :2
Actual   :0
```

**Comment by RobAustin** [comment]
regarding  https://gist.github.com/jheusser/28621f722f0afa6137e80ef0e0afc276

I believe you had a bug in your time provider
I feel rather than this :


``` java
class TestTimeProvider implements TimeProvider {

		volatile long now = System.currentTimeMillis();
		
		@Override
		public long currentTimeMillis() {
			return now;
		}
		
		public void add(long addInMs) {
			now += addInMs;
		}
		
	}
``` 

your intention may be was

``` java
    class TestTimeProvider implements TimeProvider {

        volatile long add = System.currentTimeMillis();

        @Override
        public long currentTimeMillis() {
            return System.currentTimeMillis() + add;
        }

        public void add(long addInMs) {
            add += addInMs;
        }

    }
```  

with this change, the test passes :

``` java
package net.openhft.chronicle.queue.impl.single;

import net.openhft.chronicle.core.time.TimeProvider;
import net.openhft.chronicle.queue.ExcerptAppender;
import net.openhft.chronicle.queue.ExcerptTailer;
import net.openhft.chronicle.queue.RollCycles;
import net.openhft.chronicle.queue.impl.StoreFileListener;
import net.openhft.chronicle.wire.WireType;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static org.junit.Assert.assertEquals;

public class RollCycleTest {

    Path path;

    @Before
    public void setUp() throws Exception {
        path = Files.createTempDirectory("rollCycleTest");
    }

    @After
    public void tearDown() throws IOException {
        Files.walk(path)
                .collect(Collectors.toCollection(LinkedList::new))
                .descendingIterator()
                .forEachRemaining(path -> {
                    try {
                        Files.deleteIfExists(path);
                    } catch (Exception e) {
                    }
                });
    }

    class TestTimeProvider implements TimeProvider {

        volatile long add = System.currentTimeMillis();

        @Override
        public long currentTimeMillis() {
            return System.currentTimeMillis() + add;
        }

        public void add(long addInMs) {
            add += addInMs;
        }

    }

    class ParallelQueueObserver implements Runnable, StoreFileListener {
        SingleChronicleQueue queue;
        CountDownLatch progressLatch;
        int documentsRead;

        public ParallelQueueObserver(TimeProvider timeProvider, Path path) {
            queue = SingleChronicleQueueBuilder.binary(path)
                    .rollCycle(RollCycles.DAILY).timeProvider(timeProvider).storeFileListener(this).wireType(WireType.FIELDLESS_BINARY).build();

            documentsRead = 0;
            progressLatch = new CountDownLatch(1);
        }

        @Override
        public void run() {

            ExcerptTailer tailer = queue.createTailer();

            progressLatch.countDown();

            while (!Thread.currentThread().isInterrupted()) {

                String readText = tailer.readText();
                if (readText != null) {
                    System.out.println("Read a document " + readText);
                    documentsRead++;
                }
            }
        }

        public void await() throws Exception {
            progressLatch.await();
        }

        public int documentsRead() {
            return documentsRead;
        }

        @Override
        public void onAcquired(int cycle, File file) {
            System.out.println("Acquiring " + file);
        }

        @Override
        public void onReleased(int cycle, File file) {
            System.out.println("Releasing " + file);
        }
    }

    @Test
    public void newRollCycleIgnored() throws Exception {
        TestTimeProvider timeProvider = new TestTimeProvider();
        ParallelQueueObserver observer = new ParallelQueueObserver(timeProvider, path);

        try (SingleChronicleQueue queue = SingleChronicleQueueBuilder.binary(path)
                .rollCycle(RollCycles.DAILY).timeProvider(timeProvider).wireType(WireType.FIELDLESS_BINARY).build()) {
            ExcerptAppender appender = queue.acquireAppender();

            Thread thread = new Thread(observer);
            thread.start();

            observer.await();

            // two days pass
            timeProvider.add(TimeUnit.DAYS.toMillis(2));

            appender.writeText("Day 3 data");

            // allow parallel tailer to finish iteration
            Thread.sleep(2000);

            thread.interrupt();
        }

        assertEquals(1, observer.documentsRead);
    }

    @Test
    public void newRollCycleIgnored2() throws Exception {
        TestTimeProvider timeProvider = new TestTimeProvider();
        ParallelQueueObserver observer = new ParallelQueueObserver(timeProvider, path);

        try (SingleChronicleQueue queue = SingleChronicleQueueBuilder.binary(path)
                .rollCycle(RollCycles.DAILY).timeProvider(timeProvider).wireType(WireType.FIELDLESS_BINARY).build()) {
            ExcerptAppender appender = queue.acquireAppender();
            // uncomment next line to make the test pass
            appender.writeText("Day 1 data");

            Thread thread = new Thread(observer);
            thread.start();

            observer.await();

            // two days pass
            timeProvider.add(TimeUnit.DAYS.toMillis(2));

            appender.writeText("Day 3 data");

            // allow parallel tailer to finish iteration
            Thread.sleep(2000);

            thread.interrupt();
        }

        assertEquals(2, observer.documentsRead);
    }
}
```

**Comment by RobAustin** [comment]
In addition, I have created another test that still uses two threads, using an executor.submit(), simply because with this approach you can control when the second thread is run, This is preferred over new Thread(..),start(), as it makes the test more predictable and hence easier to fix ( and debug ), but still uses two threads so will hi light any visibility issues. I've checked-in this new test based on your original work, into net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadTest. ( checked in to the snapshot )

I'm going to close this issue as ( in my view ), I feel the issue reported here, was with the test, rather than chronicle queue its self. If you feel this is not the case, please modify the net.openhft.chronicle.queue.impl.single.RollCycleMultiThreadTest and resubmit and issue along with a pull request containing the modified test, 

Regards

Rob

**Comment by jheusser** [comment]
Thanks a lot for the investigation @RobAustin . I will look into this and check the time provider and your tests.

**Comment by RobAustin** [comment]
@jheusser regarding my suggested change above to your test, strictly speaking your test should have passed as is, you should not had to change the TimeProvider,  So I have subsequently looked into this issue, and found that we did in deed, have an issue with caching of the result from time-provider this has now been fixed in the latest snapshot, we have added the following test - 

net.openhft.chronicle.queue.impl.single.SingleChronicleQueueTest#testReadWritingWithTimeProvider

Note : its unlikely to be related to the original issue that you were seeing, but jus thought you would be interested.

Regards Rob

**Comment by jheusser** [comment]
@RobAustin just FYI, I believe I have found the issue I ran into, which I caught by coincidence while having DEBUG logging on. It looked like 4.5.14 can't roll over to a new cycle under some condition and kept logging for every read on a tailer, without creating the next cycle file:
```
15:12:03.731 DEBUG n.o.c.q.i.s.SingleChronicleQueueExcerpts$StoreTailer - moveToIndex: 42d1 ffffffff
```

This seems fixed when upgrading to 4.5.19.
